src/trip_kinematics/URDFParser.py

```diff
@@ -0,0 +1,301 @@
+import xml.etree.ElementTree as ET
+from collections import defaultdict
+from typing import Dict, List
+
+import numpy as np
+from scipy.spatial.transform import Rotation as ScipyRotation
+from trip_kinematics.Transformation import Transformation
+
+
+def from_urdf(filename: str) -> List[Transformation]:
+    """Converts a robot specified in a URDF file into a list of :py:class`Transformation` objects.
+
+    If the <origin> tag does not specify xyz and rpy values or the tag is omitted, these default
+    to (0, 0, 0). <axis> defaults to (0, 0, 1).
+
+    Args:
+        filename (str): Path to URDF file.
+
+    Raises:
+        ValueError: Could not parse URDF file.
+
+    Returns:
+        List[Transformation]: List of transformations that describe the robot.
+    """
+    tree = ET.parse(filename)
+    root = tree.getroot()
+    joints = root.findall('joint')
+    joint_name_to_transformations = {}
+
+    joint_tree_dict = _build_joint_tree_dict(joints)
+
+    root_joints = [name for name, value in joint_tree_dict.items() if not value['parent']]
+
+    for i, joint in enumerate(joints):
+        name = joint.get('name')
+        if not name:
+            raise ValueError('Missing name field in URDF file')
+        joint_tree_dict[name]['joint_index'] = i
+
+    for joint in joints:
+        joint_name_to_transformations[joint.get('name')] = _get_transformations_for_joint(joint)
+
+    transformations = []
+    for joint in root_joints:
+        transformations.extend(_create_transformations_from_tree(joint,
+                                                                 joint_tree_dict,
+                                                                 joint_name_to_transformations,
+                                                                 None)
+                               )
+
+    return transformations
+
+
+def align_vectors(target: np.ndarray, to_align: np.ndarray) -> np.ndarray:
+    """Generates a rotation matrix that makes b parallel to a.
+
+    Args:
+        a (np.ndarray): 3D Vector.
+        b (np.ndarray): 3D Vector.
+
+    Returns:
+        np.ndarray: 3x3 rotation matrix.
+    """
+    target = target / np.linalg.norm(target)
+    to_align = to_align / np.linalg.norm(to_align)
+
+    # If the vectors are parallel but in opposite directions, return a 180 degree rotation
+    if np.array_equal(target, -to_align):
+        return -np.identity(3)
+
+    align_axis = np.cross(target, to_align)
+    cos_angle = np.dot(target, to_align)
+    k = 1 / (1 + cos_angle)
+
+    result = np.array([[(align_axis[0] * align_axis[0] * k) + cos_angle,

```

Comment:
This formulation obfuscates the underlying mathematical operation. I dont quite understand whats going on here.
Please either add a clarifying docstring or better yet use matrix multiplaction to better show what is happening

Reply:
Added source link in 8af1e4bd441d1528f5ec01eb04fbc889c2580ef1